pkg/proxy/iptables/proxier.go

```diff
@@ -761,6 +762,89 @@ func (proxier *Proxier) execConntrackTool(parameters ...string) error {
 	return nil
 }
 
+func parseKubeProxyTableAddRules(table utiliptables.Table, bytes []byte) (map[utiliptables.Chain][]utiliptables.Rule, error) {
+	var proxyChains []utiliptables.Chain
+	var proxyChainPrefixes []string
+	switch table {
+	case utiliptables.TableFilter:
+		proxyChains = []utiliptables.Chain{kubeServicesChain}
+	case utiliptables.TableNAT:
+		proxyChains = []utiliptables.Chain{
+			kubePostroutingChain,
+			KubeMarkMasqChain,
+			kubeServicesChain,
+			kubeNodePortsChain,
+		}
+		proxyChainPrefixes = []string{
+			"KUBE-XLB-",
+			"KUBE-SEP-",
+			"KUBE-FW-",
+			"KUBE-SVC-",
+		}
+	default:
+		return nil, fmt.Errorf("unhandled iptables table %v", table)
+	}
+
+	return utiliptables.ParseTableAddRules(table, proxyChains, proxyChainPrefixes, bytes)
+}
+
+// Returns whether the new iptables rules are the same as the old rules
+func sameIptablesRuleset(table utiliptables.Table, newLines []byte, oldLines []byte) (bool, error) {
+	oldChains, err := parseKubeProxyTableAddRules(table, oldLines)
+	if err != nil {
+		return false, err
+	}
+	newChains, err := parseKubeProxyTableAddRules(table, newLines)
+	if err != nil {
+		return false, err
+	}
```

Comment:
There's a difference between oldLines failing to parse and newLines failing to parse: oldLines can fail to parse if the administrator mucks around with iptables behind our back, or if the format of iptables-save changes. But newLines can only fail to parse in the event of a programmer error.